include/autogradpp/detail.h

```diff
@@ -33,11 +35,80 @@ extern tag::Engine engine;
 class ContainerImpl;
 class OptimizerImpl;
 using Variable = tag::Variable;
-using variable_list = tag::variable_list;
 using Tensor = at::Tensor;
 using Container = std::shared_ptr<ContainerImpl>;
 using Optimizer = std::shared_ptr<OptimizerImpl>;
 
+#define GEN_TYPE(TYP, NAME) \
+  Variant(TYP);             \
+  bool is ## NAME () const; \
+  TYP get ## NAME () const;
+
+class Variant {
+ public:
+  Variant(Tensor);
+  Variant(Variable);
+  Variant(const std::string&);
+  Variant(std::vector<Variant>&);
+  Variant(std::vector<Variant>&&);
+  Variant(std::initializer_list<Variable>);
+  Variant(std::unordered_map<std::string, Variant>&);
+  Variant(std::unordered_map<std::string, Variant>&&);
+
+  Variable const&                                       get() const;
+  std::string const&                                    getString() const;
+  std::vector<Variant> const&                           getList() const;
+  std::unordered_map<std::string, Variant> const&       getDict() const;
+  bool                                                  isVariable() const;
+  bool                                                  isString() const;
+  bool                                                  isList() const;
+  bool                                                  isDict() const;
+  GEN_TYPE(float,      Float);
+  GEN_TYPE(double,     Double);
+  GEN_TYPE(bool,       Bool);
+  GEN_TYPE(int32_t,    Int32);
+  GEN_TYPE(int64_t,    Int64);
```

Comment:
Support for unpacking from a vector templates would be nice (I think you mentioned this at some point?)

```cpp
ag::Variable v1, v2;
int32_t v3;
std::tie(v1, v2, v3) = variant.getv<ag::Variable, ag::Variable, int32_t>(); // or getn?

// Or maybe with pass-by-reference
variant.unpackTo(v1, v2, v3);
```

Reply:
Sure, this PR is already getting too big and unwieldy though, I might do it in a future PR :) Probably would name it unpackList, since it only works on std::vector<Variable>.